packages/graphql-language-service-server/src/startServer.js

```diff
@@ -8,16 +8,53 @@
  *  @flow
  */
 
+import net from 'net';
+
 import {
   processStreamMessage,
   processIPCRequestMessage,
   processIPCNotificationMessage,
 } from './MessageProcessor';
 
-export default (async function startServer(configDir: ?string): Promise<void> {
+import {
+  IPCMessageWriter,
+  SocketMessageReader,
+  SocketMessageWriter,
+} from 'vscode-jsonrpc';
+
+type ServerOptions = {
+  port?: number,
+};
+
+export default (async function startServer(
+  configDir: ?string,
+  options?: ServerOptions,
+): Promise<void> {
+  if (options && options.port) {
+    // Socket protocol support
+    const socket = net.connect(options.port);
+    socket.setEncoding('utf8');
+    const messageReader = new SocketMessageReader(socket);
+    const messageWriter = new SocketMessageWriter(socket);
+
+    messageReader.listen(message => {
+      try {
+        if (message.id != null) {
+          processIPCRequestMessage(message, configDir, messageWriter);
+        } else {
+          processIPCNotificationMessage(message, messageWriter);
+        }
+      } catch (error) {
+        // Return with error message
+        return;
```

Comment:
This comment seems off, as you're returning `undefined` here.

I'm not familiar with `SocketMessageReader`: what does it do with the return value of this callback?